[SPARK-11976][SPARKR] Support "." character in DataFrame column name

rerngvit opened the pull request:
```
## What changes were proposed in this pull request?
- Add support "." character in DataFrame column name
- Remove R code in "createDataFrame()" that replaces "." with "_"
- Remove warning suppression in createDataFrame() for iris dataset (introduced in SPARK-12034)

## How was this patch tested?
SparkR unit tests and manual testing with the R script described in SPARK-11976
```

rerngvit commented:
@sun-rui Please have a look.

AmplabJenkins commented:
Can one of the admins verify this patch?

shivaram commented:
Jenkins, ok to test

SparkQA commented:
**[Test build #62550 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62550/consoleFull)** for PR 14264 at commit [`e1cf3d5`](https://github.com/apache/spark/commit/e1cf3d5994af65a009f9c4de5d001a2e5f8bbcfe).

SparkQA commented:
**[Test build #62550 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62550/consoleFull)** for PR 14264 at commit [`e1cf3d5`](https://github.com/apache/spark/commit/e1cf3d5994af65a009f9c4de5d001a2e5f8bbcfe).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62550/
Test FAILed.

felixcheung commented:
There's this (and I thought one more, if I recall):
https://github.com/apache/spark/blob/master/R/pkg/R/DataFrame.R#L349

And its tests in test_sparkSQL.R L779 - I'm surprised these tests pass?

sun-rui commented:
@rerngvit, could you share the background that this PR can fix the issue. I see that https://issues.apache.org/jira/browse/SPARK-11976 is still open. Any other PR in Spark 2.0 make this possible?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/LogicalPlan.scala:
```diff
@@ -201,6 +201,8 @@ abstract class LogicalPlan extends QueryPlan[LogicalPlan] with Logging {
     if (!attribute.isGenerated && resolver(attribute.name, nameParts.head)) {
       Option((attribute.withName(nameParts.head), nameParts.tail.toList))
+    } else if (!attribute.isGenerated && resolver(attribute.name, nameParts.mkString("."))) {
+      Option((attribute.withName(nameParts.mkString(".")), Nil))
```
Hi, I'm just curious. Is it okay for other Spark module?
> Different from resolveAsTableColumn, this assumes `name` does NOT start with a qualifier.

rerngvit replied:
@dongjoon-hyun I do not fully understand your question. Could you please elaborate a bit more? What do you mean by "okay for other Spark module"?

SparkQA commented:
**[Test build #62612 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62612/consoleFull)** for PR 14264 at commit [`b9000c7`](https://github.com/apache/spark/commit/b9000c71be2b11e16d8749f750282ba8d4a8d248).

SparkQA commented:
**[Test build #62612 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62612/consoleFull)** for PR 14264 at commit [`b9000c7`](https://github.com/apache/spark/commit/b9000c71be2b11e16d8749f750282ba8d4a8d248).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/62612/
Test FAILed.

rerngvit commented:
@felixcheung 
I removed prohibition for "." in the function "colnames()" for SparkR in the recent patch and also its test in the file you mentioned (test_sparkSQL.R L779). Feel free to have a look.

Commit:
[SPARK-11976][SPARKR] Support "." character in DataFrame column name

    ## What changes were proposed in this pull request?
    - Add support "." character in DataFrame column name
    - Remove R code in "createDataFrame()" that replaces "." with "_"
    - Remove warning suppression in createDataFrame() for SPARK-12034
    - Remove checking for "." for function colnames()
    - Replace usage of "_" to "." for column names in test_mllib.R and test_sparkSQL.R

    ## How was this patch tested?
    SparkR unit tests and manual testing with R script described in SPARK-11976

SparkQA commented:
**[Test build #62640 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62640/consoleFull)** for PR 14264 at commit [`60d0145`](https://github.com/apache/spark/commit/60d014501661d1c400675124c99ac7896a184250).

rerngvit commented:
@felixcheung: in the recent patch, I removed checking for "." for function colnames() and its test code for the file you indicated.

rerngvit commented:
@sun-rui
>  could you share the background that this PR can fix the issue.
As stated in https://issues.apache.org/jira/browse/SPARK-11976, Spark core is already supporting column name ".".  However, it does not work without the backticks (`). The missing part to make it work is a corresponding logic in the Analyzer layer.

>  I see that https://issues.apache.org/jira/browse/SPARK-11976 is still open.
I do not fully understand your question. Could you please elaborate a bit more? **This PR** is actually SPARK-11976.

SparkQA commented:
**[Test build #62640 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/62640/consoleFull)** for PR 14264 at commit [`60d0145`](https://github.com/apache/spark/commit/60d014501661d1c400675124c99ac7896a184250).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

sun-rui commented:
@rerngvit, sorry, I mean https://issues.apache.org/jira/browse/SPARK-11977. If your PR can enable accesses to columns with "." in their names without backticks, please first submit a PR for SPARK-11977, as the change is for the Spark Core, not SparkR specific. After that PR gets merged, you can then submit a PR for SPARK-11976 which contains SparkR only changes.

rerngvit commented:
@sun-rui Thanks. I understand now. I would submit a PR to SPARK-11977 first.

sun-rui commented:
@rerngvit, I modifed the title of SPARK-11977 to a narrow scope. You can go for it.

rerngvit commented:
Since SPARK-11977 didn't get merge and this PR is blocked on that. I decided to close this PR.
